### PR TITLE
Passwordless support

### LINKED ISSUE #287: Add support for passwordless login [SDK-1940]
### Describe the problem you'd like to have solved

The SDK currently does not support passwordless login method i.e. exchanging an OTP obtained from `/passwordless/start` call for tokens.

### Describe the ideal solution

Ideally, I suppose the SDK would expose an auth API like below that supports exchanging OTP obtained via either email or sms  -
```java
AuthRequest request = auth.exchangeOtp("me@auth0.com", "123456", "email")
        .setScope("openid contacts");
try {
    TokenHolder result = request.execute();
} catch (APIException exception) {
    // api error
} catch (Auth0Exception e) {
    // request error
}
```

```java
AuthRequest request = auth.exchangeOtp("0123456789", "123456", "sms")
        .setScope("openid contacts");
try {
    TokenHolder result = request.execute();
} catch (APIException exception) {
    // api error
} catch (Auth0Exception e) {
    // request error
}
```

## Alternatives and current work-arounds

An alternative is to directly call the Auth API like any other HTTP call and not use this SDK.

### Additional context

If you are happy to accept contributions, I have a branch with changes (not complete yet) that I can tag here.

### PR DESCRIPTION
**Pull request by jimmyjames**
### Changes

Adds support for the [Passwordless APIs](https://auth0.com/docs/connections/passwordless/reference/relevant-api-endpoints). Fixes #287 

#### Details 

Three new public methods have been added to `AuthAPI`:
- `public CustomRequest<PasswordlessEmailResponse> startPasswordlessEmailFlow(String email, PasswordlessConnection connection)` - Initiate passwordless login via email.
- `public CustomRequest<PasswordlessSmsResponse> startPasswordlessSmsFlow(String phoneNumber)` - Initiate passwordless login via SMS.
- `public AuthRequest login(String emailOrPhone, PasswordlessRealm realm, char[] otp)` - Login using the passwordless grant type using the code sent to the user.

To support these new APIs, the following new models have been added:
- `PasswordlessEmailResponse` - Model for the response to the request to initiate the passwordless flow via email.
- `PasswordlessSmsResponse` - Model for the response to the request to initiate the passwordless flow via SMS.

Additionally, to model the different options options when initiating or performing passwordless login, two new enums have been added:
- `PasswordlessConnection` - The value of the type of email to send (`link` or `code`), passed as the `connection` parameter to the API.
- `PasswordlessRealm` - The realm type specified when logging in with the passwordless grant type (`email` or `code`)
Please describe both what is changing and why this is important. Include:

#### Alternatives considered and rationale

- The response models (`PasswordlessEmailResponse` and `PasswordlessSmsResponse`) _could_ be combined into one `PasswordlessResponse` model. However, since several fields would always be null depending on whether SMS or email was used, this seemed unnecessarily confusing compared to the low cost and specificity of having models specific to the type of request.
- We do not currently use enums for parameters that have a discreet set of allowable parameter values (e.g., `connection` must be either `sms` or `email`). Seems like the right place for an enum, but if we don't want to change the current pattern of the library just accepting free form strings, we can discuss further.
- The new `login` public method takes the one-time password (`otp`) as a character array, in keeping with the best practice of not using `java.util.String` for passwords. It could be argued that since the otp is only valid for one-time use, and will typically have a short duration for its validity, that we could use String. I chose to use a character array for consistency with other APIs as well as erring on the safe side.

### References

- [Using Passwordless APIs article](https://auth0.com/docs/connections/passwordless/reference/relevant-api-endpoints)
- [Passwordless API reference ](https://auth0.com/docs/api/authentication#passwordless)
- #287 

### Testing

- [X] This change adds test coverage
- [X] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All existing and new tests complete without errors

### CODE REVIEW

File: src/main/java/com/auth0/client/auth/AuthAPI.java [hunk 1]
```diff
@@ -553,7 +558,53 @@ public AuthRequest login(String emailOrUsername, char[] password, String realm)
...
+     * @param realm The realm to use.
+     * @param otp The one-time password used to authenticate using Passwordless connections
+     *
+     * @return A request to configure and execute
+     *
+     * @see <a href="https://auth0.com/docs/connections/passwordless/reference/relevant-api-endpoints">Using Passwordless APIs</a>
+     * @see com.auth0.client.auth.AuthAPI#startPasswordlessEmailFlow(String, PasswordlessEmailType)
+     * @see com.auth0.client.auth.AuthAPI#startPasswordlessSmsFlow(String)
+     */
+    public AuthRequest login(String emailOrPhone, PasswordlessRealmType realm, char[] otp) {
```
Comment (jimmyjames) [inline src/main/java/com/auth0/client/auth/AuthAPI.java:590]: We overload the `login` method several times in `AuthAPI` , so I followed that pattern here. That said, I wonder if it makes sense, or if it should be something like `exchangePasswordlessOtp` or something along those lines?

Comment (lbalmaceda) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:591]: It's a good question. I lean towards the latter because I find it easier to type and use, rather than typing "login" and scrolling through the N suggested alternatives. Also, I know we plan to add MFA as well, so that other one could look like `exchangeMfaOtp` for example. 

What's the `realm` parameter used for here? Can we assume that value instead of having to receive it? Since you have an object and then just call `getRealm` on it maybe it's simpler to receive a string. Unless you abstracted this because you are using it [on 3 or more places](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)). 

About the use of char arrays for passwords. I remember someone raised this in the past. This is a temporary code that shouldn't last more than 20 seconds. Do we still want to use char arrays? There's no wrong answer.

**Review by lbalmaceda — COMMENTED** [review]
Looks complete. Also looks like from this PR on we use custom requests and responses classes for each endpoint. I'm not saying that's wrong, on the other hand I think is a good change. But I still have some doubts about how useful some of these are when it comes to weighting simplifying the code vs adding boilerplate code.

One open question. We do have a single "login" method that accepts the "email or phone number" and the otp. But the auth is _started_ with 2 different methods, one for SMS and one for email. What are your thoughts on having a single start method that accepts the type along with the mode?

File: src/main/java/com/auth0/client/auth/AuthAPI.java [hunk 2]
```diff
@@ -34,11 +36,15 @@
     private static final String KEY_CONNECTION = "connection";
     private static final String KEY_TOKEN = "token";
     private static final String KEY_REFRESH_TOKEN = "refresh_token";
+    private static final String KEY_OTP = "otp";
+    public static final String KEY_REALM = "realm";
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:40]: No, good catch this should be private, will update.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:591]: I agree with your points regarding the name - I will change it to `exchangePasswordlessOtp`. 

> What's the realm parameter used for here? Can we assume that value instead of having to receive it? Since you have an object and then just call getRealm on it maybe it's simpler to receive a string.

The realm is used to specify the delivery type of the code sent to the user: "Use sms or email (should be the same as POST /passwordless/start)" ([docs](https://auth0.com/docs/api/authentication#authenticate-user)). Unless we create separate methods for email/sms, we cannot infer it. As to whether it should just accept a simple string, that is a viable option, and the way that other methods in this SDK work. An enum provides some additional protections as they are fixed constants, so a developer could not (for example), specify an invalid value. That said, it does also mean that if other valid realm types are added, the enum would require updating. 

> This is a temporary code that shouldn't last more than 20 seconds. Do we still want to use char arrays? There's no wrong answer.

I thought about this for the same reason (short duration code). For that reason using a String does not present significant risk, but chose a char array to err on the side of safety.

File: src/main/java/com/auth0/client/auth/AuthAPI.java [hunk 3]
```diff
@@ -698,4 +750,87 @@ public AuthRequest exchangeCode(String code, String redirectUri) {
...
+     * </pre>
+     *
+     * @param phoneNumber The phone number to send the code to. Must not be null.
+     *
+     * @return a Request to configure and execute.
+     *
+     * @see <a href="https://auth0.com/docs/connections/passwordless/guides/sms-otp">Passwordless Authentication with SMS documentation</a>
+     * @see <a href="https://auth0.com/docs/api/authentication#get-code-or-link">Get code or link API reference documentation</a>
+     */
+    public PasswordlessSmsRequest startPasswordlessSmsFlow(String phoneNumber) {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:819]: SMS only supports sending a code ([docs](https://auth0.com/docs/connections/passwordless/guides/sms-otp))

File: src/main/java/com/auth0/client/auth/PasswordlessRealmType.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    EMAIL("email"),
+
+    /**
+     * The sms realm.
+     */
+    SMS("sms");
+
+    private final String realm;
+
+    PasswordlessRealmType(String realm) {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/PasswordlessRealmType.java:20]: enums cannot have public or protected constructors (and their visibility is private by default). [SO question](https://stackoverflow.com/questions/3664077/why-cant-enum-constructors-be-protected-or-public-in-java) with some context and links.

File: src/main/java/com/auth0/client/auth/PasswordlessEmailType.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package com.auth0.client.auth;
+
+/**
+ * Represents the type of the Passwordless email request.
+ */
+public enum PasswordlessEmailType {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/PasswordlessEmailType.java:6]: How about `PasswordlessEmailConnection`, since it represents the value of the `connection` parameter for email ([docs](https://auth0.com/docs/api/authentication#get-code-or-link))?

File: src/main/java/com/auth0/net/PasswordlessEmailRequest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.Map;
+
+/**
+ * Represents a request to send a code or link via email to begin the authentication flow for passwordless connections.
+ *
+ * @see <a href="https://auth0.com/docs/connections/passwordless/guides/email-otp">Passwordless Authentication with Email documentation</a>
+ * @see <a href="https://auth0.com/docs/api/authentication#get-code-or-link">Get code or link API reference documentation</a>
+ */
+public class PasswordlessEmailRequest extends CustomRequest<PasswordlessEmailResponse> {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/net/PasswordlessEmailRequest.java:15]: The passwordless email request allow specifying additional `authParams` - "Use this to append or override the link parameters (like scope, redirect_uri, protocol, response_type), when you send a link using email." ([docs](https://auth0.com/docs/api/authentication#get-code-or-link)). If we use `CustomRequest<PasswordlessEmailResponse>`, we'd need to add the authParams as a parameter to the methods (and either allow but require `null`, or overload the methods so as not to require developers to pass null when they don't need too specify those params). 

I think it's ok to not add these request classes, and instead allow passing the `authParams` to the `startPasswordlessEmailFlow` method (and I would advocate for adding an overloaded method so developers aren't forced to pass `null` when it's not needed or not applicable (if not sending a link)).

**Comment by jimmyjames** [comment]
> Looks complete. Also looks like from this PR on we use custom requests and responses classes for each endpoint. I'm not saying that's wrong, on the other hand I think is a good change. But I still have some doubts about how useful some of these are when it comes to weighting simplifying the code vs adding boilerplate code.

We can implement this without the new request types as discussed above. We do need new classes to represent the responses.
 
> One open question. We do have a single "login" method that accepts the "email or phone number" and the otp. But the auth is _started_ with 2 different methods, one for SMS and one for email. What are your thoughts on having a single start method that accepts the type along with the mode?

Having one start method would probably be possible, but IMO would have poor usability since we'd need to expose several parameters that are only valid for one type of connection.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java]: Updated to private.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 1 above)
Comment (lbalmaceda) [reply src/main/java/com/auth0/client/auth/AuthAPI.java]: re char array / string, I'm fine either way. re realm, do notice that passwordless connections other than the default "SMS" and "email" **can** be created via the management API. So we should support any value here.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/main/java/com/auth0/client/auth/AuthAPI.java]: They must have changed this, 4 years ago it was supported to use "link". I think it's enough to support "code".  Especially when you think SMS is limited in their length.

File: src/main/java/com/auth0/client/auth/AuthAPI.java [hunk 7]
```diff
@@ -698,4 +750,89 @@ public AuthRequest exchangeCode(String code, String redirectUri) {
...
+     *
+     * @param email the email address to send the code or link to. Must not be null.
+     * @param connection the type of the passwordless email request. Must not be null.
+     *
+     * @return a Request to configure and execute.
+     *
+     * @see <a href="https://auth0.com/docs/connections/passwordless/guides/email-otp">Passwordless Authentication with Email documentation</a>
+     * @see <a href="https://auth0.com/docs/api/authentication#get-code-or-link">Get code or link API reference documentation</a>
+     */
+    public CustomRequest<PasswordlessEmailResponse> startPasswordlessEmailFlow(String email, PasswordlessEmailConnection connection) {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/auth/AuthAPI.java:777]: I still find it a bit misleading the class name "PasswordlessEmailConnection". Even the param name "connection". This is not a connection. The API defines it as "send" or "what to send to the user". And values can be link or code. I think using something like "mode" or "type" makes more sense. If we use connection it could be interpreted as "realm", and that's not what this is.

File: src/main/java/com/auth0/client/auth/AuthAPI.java [hunk 8]
```diff
@@ -698,4 +750,89 @@ public AuthRequest exchangeCode(String code, String redirectUri) {
...
+                .addPathSegment(PATH_PASSWORDLESS)
+                .addPathSegment(PATH_START)
+                .build()
+                .toString();
+
+        CustomRequest<PasswordlessEmailResponse> request = new CustomRequest<>(client, url, "POST", new TypeReference<PasswordlessEmailResponse>() {
+        });
+        request.addParameter(KEY_CLIENT_ID, clientId);
+        request.addParameter(KEY_CLIENT_SECRET, clientSecret);
+        request.addParameter(KEY_CONNECTION, "email");
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/auth/AuthAPI.java:792]: Do we want to extract this "connection" value to the method signature? (without adding another overloaded method) or should we leave that edge case (pwdless connections created via the API) to be updated later via the "set parameters" method that every CustomRequest has?

File: src/main/java/com/auth0/client/auth/PasswordlessRealm.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package com.auth0.client.auth;
+
+/**
+ * Represents the type of realm used for Passwordless authentication.
+ */
+public enum PasswordlessRealm {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/auth/PasswordlessRealm.java:6]: the only reason to not use enums is that they can't be instantiated outside this class, can they? so there's no support for custom passwordless connections. 

Ideally, we could have this good defaults and then let the developer do something like this

```java
new PasswordlessRealm("sms-custom");
```

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 8 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:792]: I'm not sure I follow. This method is specific to starting the password flow using email, so the connection value wouldn't change.

File: src/main/java/com/auth0/client/auth/PasswordlessRealm.java (hunk 9 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/PasswordlessRealm.java:6]: Right, enums don't really support the notion of being instantiated - they represent a fixed set of values. Since an enum wouldn't support custom passwordless connections, I'll update to use a String param in the API and remove this enum.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 8 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:792]: Ah, I understand now, you are referring to the ability to create custom passwordless connections via the Management API. I'll look into the best way to handle this; on first glance I'm inclined to support this via overriding the parameters via `CustomRequest` for this edge case.

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 7 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:777]: Sure, I'll change it back to `PasswordlessEmailType` as it was before. I changed it here based on an earlier review comment, but I Iagree that `connection` is misleading, even though that is the parameter name in the API.

File: README.md [hunk 10]
```diff
@@ -245,6 +245,42 @@ try {
...
+This library supports [Passwordless Authentication](https://auth0.com/docs/connections/passwordless) to allow users to log in without the need to remember a password.
+
+The email flow supports sending both a code or link to initiate login:
+
+```java
+try {
+    PasswordlessEmailResponse = auth.startPasswordlessEmailFlow("user@domain.com", PasswordlessEmailType.CODE)
+        .execute();
+} catch (Auth0Exception e) {
+        // handle request error
```
Comment (lbalmaceda) [inline README.md:259]: nit, white-spacing should match the rest

File: src/main/java/com/auth0/client/auth/AuthAPI.java (hunk 8 above)
Comment (lbalmaceda) [reply src/main/java/com/auth0/client/auth/AuthAPI.java:792]: >  on first glance I'm inclined to support this via overriding the parameters via CustomRequest for this edge case.
I think that's fine, given that the creation of those connections is not straightforward from the dashboard.

File: src/test/java/com/auth0/client/auth/AuthAPITest.java [hunk 11]
```diff
@@ -944,6 +939,171 @@ public void shouldCreateLogInWithClientCredentialsGrantRequest() throws Exceptio
...
+
+    @Test
+    public void startPasswordEmailFlowShouldThrowWhenTypeIsNull() throws Exception {
+        exception.expect(IllegalArgumentException.class);
+        exception.expectMessage("'type' cannot be null!");
+        Request<PasswordlessEmailResponse> request = api.startPasswordlessEmailFlow("user@domain.com", null);
+    }
+
+    @Test
+    public void shouldCreaateStartEmailPasswordlessFlowRequestWithCustomParams() throws Exception {
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/auth/AuthAPITest.java:984]:
```suggestion
    public void shouldCreateStartEmailPasswordlessFlowRequestWithCustomParams() throws Exception {
```
typo

File: src/test/java/com/auth0/client/auth/AuthAPITest.java [hunk 12]
```diff
@@ -944,6 +939,171 @@ public void shouldCreateLogInWithClientCredentialsGrantRequest() throws Exceptio
         assertThat(response.getExpiresIn(), is(notNullValue()));
     }
 
+    // Login with Passwordless
+
+    @Test
+    public void shouldCreaateStartEmailPasswordlessFlowRequest() throws Exception {
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/auth/AuthAPITest.java:945]:
```suggestion
    public void shouldCreateStartEmailPasswordlessFlowRequest() throws Exception {
```

File: src/test/java/com/auth0/client/auth/AuthAPITest.java [hunk 13]
```diff
@@ -944,6 +939,171 @@ public void shouldCreateLogInWithClientCredentialsGrantRequest() throws Exceptio
...
+        assertThat(body, hasEntry("email", (Object) "user@domain.com"));
+
+        assertThat(response, is(notNullValue()));
+        assertThat(response.getEmail(), not(isEmptyOrNullString()));
+        assertThat(response.getId(), not(isEmptyOrNullString()));
+        assertThat(response.isEmailVerified(), is(notNullValue()));
+    }
+
+    @Test
+    public void startPasswordEmailFlowShouldThrowWhenEmailIsNull() throws Exception {
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/auth/AuthAPITest.java:970]: should these be "passwordless" instead of "password"? repeats on the others
```suggestion
    public void startPasswordlessEmailFlowShouldThrowWhenEmailIsNull() throws Exception {
```

File: src/test/java/com/auth0/client/auth/AuthAPITest.java [hunk 14]
```diff
@@ -944,6 +939,171 @@ public void shouldCreateLogInWithClientCredentialsGrantRequest() throws Exceptio
...
+        exception.expect(IllegalArgumentException.class);
+        exception.expectMessage("'email' cannot be null!");
+        Request<PasswordlessEmailResponse> request = api.startPasswordlessEmailFlow(null, PasswordlessEmailType.CODE);
+    }
+
+    @Test
+    public void startPasswordEmailFlowShouldThrowWhenTypeIsNull() throws Exception {
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/auth/AuthAPITest.java:977]:
```suggestion
    public void startPasswordlessEmailFlowShouldThrowWhenTypeIsNull() throws Exception {
```

**Review by lbalmaceda — APPROVED** [review]
Looks good so I've approved, a few typos left.

File: README.md (hunk 10 above)
Comment (jimmyjames) [reply README.md:259]: Pushed a commit to fix

**Review by lbalmaceda — APPROVED** [review]
